src/common/prometheus/prometheus.service.ts

```diff
@@ -383,5 +374,38 @@ export class PrometheusService implements OnApplicationBootstrap {
         throw e;
       })
       .finally(() => stop());
-  }
+  };
+}
+
+export function TrackTask(name: string) {
+  return (target: any, propertyKey: string, descriptor: PropertyDescriptor) => {
+    const originalValue = descriptor.value;
+
+    descriptor.value = function (...args) {
+      // "this" here will refer to the class instance
+      if (!this.prometheus) throw Error(`'${this.constructor.name}' class object must contain 'prometheus' property`);
+      const stop = this.prometheus.taskDuration.startTimer({
+        name: name,
+      });
+      this.logger.debug(`Task '${name}' in progress`);
+      return originalValue
+        .apply(this, args)
+        .then((r) => {
+          this.prometheus.taskCount.inc({
+            name: name,
+            status: TaskStatus.COMPLETE,
+          });
+          return r;
+        })
+        .catch((e) => {
+          this.logger.error(`Task '${name}' ended with an error`, e.stack);
+          this.prometheus.taskCount.inc({
+            name: name,
+            status: TaskStatus.ERROR,
+          });
+          throw e;
+        })
+        .finally(() => this.logger.debug(`Task '${name}' is complete. Duration: ${stop()}`));
```

Comment:
Extract `stop()` call from the template string to be explicit.

Reply:
done